File: linux/bug.py
import math
import random


class BugState:
    WALKING  = 'walking'
    IDLE     = 'idle'
    SQUISHED = 'squished'
    SPLAT    = 'splat'


class Bug:
    HIT_RADIUS      = 20   # direct squish
    NEAR_MISS_RADIUS = 55  # triggers miss sound

    def __init__(self, screen_w, screen_h):
        self._w = screen_w
        self._h = screen_h
        margin = 80
        self.x     = random.uniform(margin, screen_w - margin)
        self.y     = random.uniform(margin, screen_h - margin)
        self.angle = random.uniform(0, 2 * math.pi)
        self.speed = random.uniform(80, 145)
        self.state       = BugState.WALKING
        self.state_timer = random.uniform(1.0, 3.5)
        self.leg_phase   = 0.0
        self.splat_shape = []   # pre-computed polygon points (local coords)
        self.splat_alpha = 1.0

    # ------------------------------------------------------------------
    # Update

    def update(self, dt):
        self.state_timer -= dt

        if self.state == BugState.WALKING:
            self.leg_phase = (self.leg_phase + dt * 10) % 1.0
            self.x += math.cos(self.angle) * self.speed * dt
            self.y += math.sin(self.angle) * self.speed * dt
            self._bounce()
            if self.state_timer <= 0:
                if random.random() < 0.28:
                    self.state       = BugState.IDLE
                    self.state_timer = random.uniform(0.3, 1.4)
                else:
                    self.angle      += random.uniform(-1.1, 1.1)
                    self.state_timer = random.uniform(0.8, 3.0)

        elif self.state == BugState.IDLE:
            if self.state_timer <= 0:
                self.state       = BugState.WALKING
                self.state_timer = random.uniform(1.0, 3.5)

        elif self.state == BugState.SPLAT:
            self.splat_alpha -= dt * 0.55
            if self.splat_alpha < 0:
                self.splat_alpha = 0

    # ------------------------------------------------------------------
    # State transitions

    def squish(self):
        self.state = BugState.SQUISHED
        self._build_splat_shape()

    def become_splat(self):
        self.state       = BugState.SPLAT
        self.splat_alpha = 1.0

    def respawn(self):
        margin = 80
        self.x     = random.uniform(margin, self._w - margin)
        self.y     = random.uniform(margin, self._h - margin)
        self.angle = random.uniform(0, 2 * math.pi)
        self.speed = random.uniform(80, 145)
        self.state       = BugState.WALKING
        self.state_timer = random.uniform(1.0, 3.0)
        self.leg_phase   = 0.0
        self.splat_alpha = 1.0

    def is_alive(self):
        return self.state in (BugState.WALKING, BugState.IDLE)

    # ------------------------------------------------------------------
    # Internals

    def _bounce(self):
        m = 30
        if self.x < m:
            self.x     = m
            self.angle = math.pi - self.angle
        elif self.x > self._w - m:
            self.x     = self._w - m
            self.angle = math.pi - self.angle
        if self.y < m:
            self.y     = m
            self.angle = -self.angle
        elif self.y > self._h - m:
            self.y     = self._h - m
            self.angle = -self.angle

    def _build_splat_shape(self):
        """Pre-compute blob polygon so it's stable across frames."""
        self.splat_shape = []
        n = 14
        for i in range(n):
            a = i * 2 * math.pi / n
            r = random.uniform(14, 22) if i % 2 == 0 else random.uniform(7, 12)
            self.splat_shape.append((math.cos(a) * r, math.sin(a) * r))

File: mac/Sources/Bug.swift
import AppKit

enum BugState: Equatable {
    case walking
    case idle
    case turning
    case squished
    case splat
}

class Bug {
    var position: CGPoint
    var angle: CGFloat        // radians; 0 = facing right (+x)
    var speed: CGFloat        // pixels per second
    var state: BugState = .walking
    var stateTimer: TimeInterval = 0
    var legPhase: CGFloat = 0 // 0.0–1.0 cycling, drives leg swing animation
    var splatShape: [CGPoint] = []
    var splatAlpha: CGFloat = 1.0

    // Tunable radii
    let hitRadius: CGFloat = 20
    let nearMissRadius: CGFloat = 55

    init(on screen: NSRect) {
        let m: CGFloat = 80
        position = CGPoint(
            x: .random(in: m...(screen.width - m)),
            y: .random(in: m...(screen.height - m))
        )
        angle = .random(in: 0...(2 * .pi))
        speed = .random(in: 80...145)
        stateTimer = .random(in: 1.0...3.5)
    }

    func update(dt: TimeInterval, in screen: NSRect) {
        stateTimer -= dt

        switch state {
        case .walking:
            legPhase += CGFloat(dt) * 10
            if legPhase > 1 { legPhase -= 1 }

            position.x += cos(angle) * speed * CGFloat(dt)
            position.y += sin(angle) * speed * CGFloat(dt)
            bounceEdges(in: screen)

            if stateTimer <= 0 {
                if Double.random(in: 0...1) < 0.28 {
                    state = .idle
                    stateTimer = .random(in: 0.3...1.4)
                } else {
                    angle += .random(in: -1.1...1.1)
                    stateTimer = .random(in: 0.8...3.0)
                }
            }

        case .idle:
            if stateTimer <= 0 {
                state = .walking
                stateTimer = .random(in: 1.0...3.5)
            }

        case .turning:
            // reserved for future smooth-turn AI upgrade
            state = .walking
            stateTimer = .random(in: 1.0...3.0)

        case .squished:
            break  // BugController drives transition → splat

        case .splat:
            splatAlpha -= CGFloat(dt) * 0.55
            if splatAlpha < 0 { splatAlpha = 0 }
        }
    }

    func squish() {
        state = .squished
        buildSplatShape()
    }

    func becomeSplat() {
        state = .splat
        splatAlpha = 1.0
    }

    func respawn(on screen: NSRect) {
        let m: CGFloat = 80
        position = CGPoint(
            x: .random(in: m...(screen.width - m)),
            y: .random(in: m...(screen.height - m))
        )
        angle = .random(in: 0...(2 * .pi))
        speed = .random(in: 80...145)
        state = .walking
        stateTimer = .random(in: 1.0...3.0)
        legPhase = 0
        splatAlpha = 1.0
    }

    private func bounceEdges(in screen: NSRect) {
        let m: CGFloat = 30
        if position.x < m          { position.x = m;               angle = .pi - angle }
        if position.x > screen.width  - m { position.x = screen.width  - m; angle = .pi - angle }
        if position.y < m          { position.y = m;               angle = -angle }
        if position.y > screen.height - m { position.y = screen.height - m; angle = -angle }
    }

    // Pre-compute splat polygon so it's stable across frames (no jitter)
    private func buildSplatShape() {
        splatShape = []
        let numSpikes = 14
        for i in 0..<numSpikes {
            let a = CGFloat(i) * 2 * .pi / CGFloat(numSpikes)
            // Alternate long/short spikes for blobby look
            let r: CGFloat = i % 2 == 0
                ? .random(in: 14...22)
                : .random(in: 7...12)
            splatShape.append(CGPoint(x: cos(a) * r, y: sin(a) * r))
        }
    }
}
